### PR TITLE
Remove if-else and torch.tensor to meet cudagraph requirement

### PR DESCRIPTION
**Pull request by katec846**
# Description

Please include a brief summary of the changes, relevant motivation and context.

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Remove the if-else statement for enabling cuda graph

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by ksivaman** [comment]
@sanandaraj5597 Could you take a look?

**Comment by sanandaraj5597** [comment]
This kinda removes a tiny optimization. Which is not compatible with CG?
1) If-else or 2) torch.Tensor?

**Comment by jiemingz** [comment]
`if torch.equal(grad_output, torch.tensor(1.0, device=grad_output.device))`

@sanandaraj5597 the `if` causes a device host sync which is not cudagraphable, because the data is on gpu but the 'if' is executed by host

@katec846 can we guard this check with a flag?

**Comment by sanandaraj5597** [comment]
I see, we never hit a True on this `if` condition in MCore cases. But let's add a `is_cg_capturable` flag from top to CE loss function and guard this `if` statement under that?

We can default it to False, maybe can pass it to True for the MCore case when we use full iteration CG.

**Comment by katec846** [comment]
@sanandaraj5597 can you help review? I don't have permission to add reviewers

**Comment by sanandaraj5597** [comment]
@katec846 Have you turned on CG and matched training loss curves?

**Comment by katec846** [comment]
I've compared the loss with CG on and off with llama8b finetuning, the loss are almost the same for first 20 iters. Do I need to run more iters?

**Comment by sanandaraj5597** [comment]
Thanks for comparing. LGTM! @ksivaman we can merge.

**Review by ksivaman — APPROVED** [review]
